src/iop/hazeremoval.c

```diff
@@ -0,0 +1,973 @@
+/*
+    This file is part of darktable,
+    copyright (c) 2017 Heiko Bauke.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+    This module implements automatic single-image haze removal as
+    described by K. He et al. in
+
+    * Kaiming He, Jian Sun, and Xiaoou Tang, "Single Image Haze
+      Removal Using Dark Channel Prior," IEEE Transactions on Pattern
+      Analysis and Machine Intelligence, vol. 33, no. 12,
+      pp. 2341-2353, Dec. 2011. DOI: 10.1109/TPAMI.2010.168
+
+    * K. He, J. Sun, and X. Tang, "Guided Image Filtering," Lecture
+      Notes in Computer Science, pp. 1-14, 2010. DOI:
+      10.1007/978-3-642-15549-9_1
+*/
+
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "common/darktable.h"
+#include "bauhaus/bauhaus.h"
+#include "develop/imageop.h"
+#include "develop/tiling.h"
+#include "gui/gtk.h"
+#include "iop/iop_api.h"
+
+#include <gtk/gtk.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+#include <float.h>
+
+//----------------------------------------------------------------------
+// implement the module api
+//----------------------------------------------------------------------
+
+DT_MODULE_INTROSPECTION(1, dt_iop_hazeremoval_params_t)
+
+typedef float rgb_pixel[3];
+
+typedef struct dt_iop_hazeremoval_params_t
+{
+  float strength;
+  float distance;
+} dt_iop_hazeremoval_params_t;
+
+typedef struct dt_iop_hazeremoval_data_t
```

Comment:
just typedef `dt_iop_hazeremoval_data_t` to be `dt_iop_hazeremoval_params_t`, and drop `commit_params()`